### ISSUE TITLE
In some cases sorting using String.casecmp results in Comparison method violates its general contract!

### ISSUE THREAD

**Issue by mrckzgl** [jruby/jruby#7946]
This is a very strange issue, only happening in jruby (for regular ruby it does not).
Consider this minimal example:

We got some json files with text values we want to sort:
```
require "json/ext"

values = files.map{|f|
  plain_text=File.read(f);
  json = JSON.parse(plain_text, :symbolize_names => true)
  json[:text][0..0]
}

values_copied = values.join("\n").split("\n")
```
In the real world we sort the whole text, but for this example single characters are enough.
The [list](https://github.com/jruby/jruby/files/12803495/list.txt) contains 262 character values, a lot of them are duplicates (not sure if this matters). Characters are latin, arabic and '['.

The problem is, that this list is not sortable using `casecmp`:
```
values.sort!{|a,b|
  a.casecmp(b)
}
```
will result in
```
Unhandled Java exception: java.lang.IllegalArgumentException: Comparison method violates its general contract!
java.lang.IllegalArgumentException: Comparison method violates its general contract!
              mergeHi at java/util/TimSort.java:903
              mergeAt at java/util/TimSort.java:520
        mergeCollapse at java/util/TimSort.java:448
                 sort at java/util/TimSort.java:245
                 sort at java/util/Arrays.java:1306
         sortInternal at org/jruby/RubyArray.java:4138
            sort_bang at org/jruby/RubyArray.java:3988
                 call at org/jruby/RubyArray$INVOKER$i$0$0$sort_bang.gen:-1
[...]
```

Now for the strange parts.

1) If we sort `values_copied` instead of `values` directly, sorting works without a problem.
2) If we extract the first character via `json[:text][0]` instead of `json[:text][0..0]` it also works.
3) Also saving the terminal output of `puts(values)` into a file (like the attached), load that list via `split`, sort also works.
4) Case sensitive comparison using `<=>` also works without a problem.

So, I thought it might have something to do with the internal byte representation of the values.
However, if I compare the byte representation of the elements of `values` with `values_copied` via `unpack("C*")` they are identical.

Does anybody have an idea, what could cause the exception?

**Environment Information**

`jruby 9.4.1.0 (3.1.0) 2023-02-07 237d5fa5f4 Java HotSpot(TM) 64-Bit Server VM 15.0.2+7-27 on 15.0.2+7-27 +jit [x86_64-linux]`
`Ubuntu 22.04 / 6.2.0-33-generic #33~22.04.1-Ubuntu SMP PREEMPT_DYNAMIC Thu Sep  7 10:33:52 UTC 2 x86_64 x86_64 x86_64 GNU/Linux`

**Comment by headius** [comment]
It seems this might indicate a problem in the comparison operator used for the sort; something about the pre-copy list is triggering it.

https://stackoverflow.com/questions/11441666/java-error-comparison-method-violates-its-general-contract/11441813#11441813

Are you able to turn this into a simple reproduction, perhaps as a git repository? I'd like to reproduce it locally.

**Comment by derMart** [comment]
Yes, this is the best I can boil it down to: [casecmp_bug.zip](https://github.com/jruby/jruby/files/12810000/casecmp_bug.zip)

Just run the .rb file with jruby 9.4.1.0. Also interesing is, that the data files text properties need to include a certain amount of characters for the exception to appear. If I just export 1 or even 10 characters from the text, it does not. Even though the ruby code extracts only the first occuring character. But maybe [0..0] in ruby behaves differently than [0] substring access in node.js which I used to extract the data. Just thinking about ltr vs. rtl direction (though I wouldn't expect substring access to care...). By the way I do think the arabic strings might be the culprit here.

Good Luck!

best

**Comment by mrckzgl** [comment]
Yes, this is the best I can boil it down to: [casecmp_bug.zip](https://github.com/jruby/jruby/files/12810000/casecmp_bug.zip)

Just run the .rb file with jruby 9.4.1.0. Also interesing is, that the data files text properties need to include a certain amount of characters for the exception to appear. If I just export 1 or even 10 characters from the text, it does not. Even though the ruby code extracts only the first occuring character. But maybe [0..0] in ruby behaves differently than [0] substring access in node.js which I used to extract the data. Just thinking about ltr vs. rtl direction (though I wouldn't expect substring access to care...). By the way I do think the arabic strings might be the culprit here.

Good Luck!

best

**Comment by headius** [comment]
> I do think the arabic strings might be the culprit here

This is a good theory, especially if the way we are comparing them might not be transitive.

I'll take a look at your example and see if I can make some progress.

**Comment by headius** [comment]
It seems like 256 elements might be the magic number after which it breaks, which could simply be an internal chunk size in JDK's Timsort. I'm still investigating why it seems to fail at this specific limit.

FWIW a workaround in the short term is to set the JVM to use the old mergesort using the JVM property "java.util.Arrays.useLegacyMergeSort" to "true", either using the `-D` JVM flag or in JRuby with `-J-Djava.util.Arrays.useLegacyMergeSort=true`.

**Comment by headius** [comment]
Ok this appears to be an issue with calculating the code range for the string; it is getting marked as CR_7BIT which is throwing off the comparison.

If print out the value of both strings inside your `sort!` comparator, it works fine.

If I remove logic [here](https://github.com/jruby/jruby/blob/373dc0b835616792fb2d8e8772e7aef0b207ee76/core/src/main/java/org/jruby/RubyString.java#L1669-L1671) that attempts to do a fast single-byte comparison, it also works properly (the multi-byte logic does not trigger an error).

So I suspect that printing the strings out is forcing a recalculation of their code range, which corrects the bug.

The bug to fix is possibly somewhere in the building of the strings, either during the JSON parse or within JRuby when we calculate the code range as the strings are gathered.

**Comment by headius** [comment]
@mrckzgl You may be on to something with your theory about `[0..0]` being part of the problem. If I replace that with `[0]` the issue also goes away.

It could be there's a bug in how we are handling substrings for ranges that's messing with the String code range. cc @enebo

**Comment by headius** [comment]
The following patch fixes it by skipping Range-based substring logic when the length would be 1:

```diff
diff --git a/core/src/main/java/org/jruby/RubyString.java b/core/src/main/java/org/jruby/RubyString.java
index 154d236076..cca1475647 100644
--- a/core/src/main/java/org/jruby/RubyString.java
+++ b/core/src/main/java/org/jruby/RubyString.java
@@ -3796,6 +3796,7 @@ public class RubyString extends RubyObject implements CharSequence, EncodingCapa
         } else if (arg instanceof RubyRange) {
             int len = strLength();
             int[] begLen = ((RubyRange) arg).begLenInt(len, 0);
+            if (begLen[1] == 1) return op_aref(runtime, begLen[0]);
             return begLen == null ? context.nil : substr19(runtime, begLen[0], begLen[1]);
         } else {
             StringSites sites = sites(context);
```

However, this does not fix whatever is broken inside the Range logic.

**Comment by headius** [comment]
I've pushed a possible fix in #7950 but I do not have a good explanation for why it fixes the problem.

**Comment by enebo** [comment]
@headius heh...upper case?  I am still boggled by the notion this would fail because of cr.  I am guessing perhaps the upper case converts the char from being single-byte perhaps (is that really possible?).  If so we say "ah single byte chars" then convert to something which isn't for the comparison.  Just thinking outloud.

**Comment by headius** [comment]
@enebo Tracking the `str[0..0]` vs `str[0]` logic the only difference I saw was that `s[0..0]` forces a `strLength` calculation which populates CR. Based on that being "valid" different paths are taken during the `substr19` logic, but I did not see *why* the difference mattered.

**Comment by headius** [comment]
This seems to be "fixed" by #7950 but we still have not found a good explanation for the problem. It may simply be that doing a case-insensitive match with these strings is not transitive, or there may be something else we're doing wrong with the constructed strings that makes them casecmp intransitively.

I'll chat with @enebo about whether we should keep this open or have a new bug or just pretend it's ok.

**Comment by enebo** [comment]
@headius I think we should examine the test case zip and see if we can figure this out.  CR of VALID fails where CR of 7BIT succeeds.  Both should work but obviously but VALID is the slow path.  The upcase vs downcase is likely doing something no one (of us at least) expects.  My speculation above is all I can think of.
